src/analysis_and_optimization/dune

```diff
@@ -1,7 +1,7 @@
 (library
  (name analysis_and_optimization)
  (public_name stanc.analysis)
- (libraries core_kernel str fmt common middle)
+ (libraries core_kernel str fmt common middle frontend)
```

Comment:
It depends on Environment.ml

Reply:
Could that dependency go in the other direction? I mean, make `SignatureMismatch.matching_function` into `Environment.matching_function` and now it's Environment that depends on SignatureMismatch.

Reply:
We could do that, but `Environment` depends on `Stan_math_signatures` in the same way `SignatureMismatch` does, so I don't think that actually buys us anything in terms of the original question

Reply:
`Stan_math_signatures` lives in `Middle` so there's no dependency on `Frontend`, which was the original question I think.

Reply:
SignatureMismatch and Promotion can be moved to Middle
https://github.com/WardBrian/stanc3/compare/dedupe-unsized-signaturemismatch...nhuurre:review1/dedupe-unsized-signaturemismatch
Changing `Debug_data_generation` to use the MIR instead of AST is possible but requires a lot more work.

Reply:
Ah, that's a fair perspective. A long term goal of mine is to make the typechecker less dependent on Stan_math_signatures, which I think leads to a different organization than your proposal. I think a big barrier to alternative backends is the requirement that they expose the same set of functions right from the start. 
I've talked to people like those working on https://github.com/deepppl and a huge help for them would be if the typechecker segment of the frontend was parameterized over the `Stan_math_signatures` module, so that they could provide an alternative set of signatures very easily, and then their version of the compiler would fail (at typechecking time) if users used unsupported or not-yet-implemented functions. This is still sort of far off IMO, but I've been trying to slowly decouple them in reasonable ways.

I agree, Promotion.ml exists purely for inter-typechecking communication. The alternative is that functions like matching_function could accept the argument list (rather than just it's types) and return a promoted list (rather than a list of promotions) - but that seemed too tightly coupled for my taste, and not doing that also came in handy when doing #1102 


Re: The partial evaluator: I agree. I think it's better to fail noisily if it does something wrong. As far as I can tell that logic has been around since #70, which started all the optimization work. IMO, some of this discussion and that change might be better suited for a separate PR?

Reply:
The signature mismatch types don't have anything specific to the typechecker but on reflection I think you're right that SignatureMismatch is mostly a submodule of Semantic_error. So yes, it should stay in Frontend. Partial evaluator can be decoupled from Stan_math_signatures/whatever later.

Moving Debug_data_generation seems like "sneaking in" an unrelated change, though. If you want it out of frontend then you can rewrite it to use MIR instead of AST but that's a different PR.

Reply:
Debug_data, to me makes sense either in the frontend (it operates on the AST, is directly called by the stanc3 binary or js interface, etc) or in analysis (it is a form of analysis, IMO). 

The difficulty is it depends on the partial evaluator. This means that it staying in frontend creates a circular dependency, since the partial evaluator (again, perhaps for poor reasons) depends on type checking, and the debug data depends on partial eval. 

Moving it was the least invasive way to fix this, and I was able to justify it to myself, but I feel much less strongly about this than the other questions of what belongs where. 

If you'd like I can try to keep it in frontend and see if I can break the loop another way, or I can try to convert it to MIR (I think this won't be too bad?)

Reply:
Aha, it's related through a dependency cycle! That makes sense.
Analysis is a natural place for datagen so no need to keep it in frontend and converting it to MIR-only is not urgent.